jetbrains-core/src/software/aws/toolkits/jetbrains/utils/FileInfoCache.kt

```diff
@@ -3,35 +3,104 @@
+     */
+    fun evaluate(path: String): Promise<T> {
+        logger.info { "Evaluating $path" }
+
+        val asyncPromise = AsyncPromise<T>()
+        asyncPromise
+            .onSuccess { result ->
+                logger.info { "File info evaluation is completed: '$result'" }
+            }
+            .onError { error -> // Need to set an error handler early, else the setError call will throw AssertionError
+                if (error !is ProcessCanceledException) {
+                    logger.info(error) { "Failed to evaluate $path" }
+                }
+            }
+
+        val infoResult = synchronized(lock) {
+            getCacheEntry(path, asyncPromise)
+        }
+
+        if (infoResult.result == asyncPromise && asyncPromise.isPending) {
+            ApplicationManager.getApplication().executeOnPooledThread {
+                try {
+                    val result = getFileInfo(path)
+                    asyncPromise.setResult(result)
+                } catch (t: Throwable) {
+                    asyncPromise.setError(t)
+                }
+            }
+        }
+
+        return infoResult.result
+    }
+
+    private fun getCacheEntry(path: String, asyncPromise: AsyncPromise<T>): InfoResult<T> {
+        val cacheEntry = infoCache[path]
 
-    @Synchronized
-    fun getResult(path: String): T {
-        val cacheResult = infoCache.remove(path)
         val currentLastModificationDate = try {
             Files.getLastModifiedTime(Paths.get(path)).toInstant()
         } catch (e: NoSuchFileException) {
-            throw IllegalStateException(message("general.file_not_found", path))
+            // If unable to get the current time, override the cache entry that the file can't be found
+
+            asyncPromise.setError(IllegalStateException(message("general.file_not_found", path)))
+
+            val newResult = InfoResult(asyncPromise, Instant.MIN)
+            infoCache[path] = newResult
+            return newResult
         }
 
-        val infoResult = if (cacheResult == null || currentLastModificationDate.isAfter(cacheResult.timestamp)) {
-            InfoResult(getFileInfo(path), currentLastModificationDate)
+        // If the promise is fulfilled, and the path has been modified since last checking, we need to check again
+        return if (cacheEntry == null ||
+            (!cacheEntry.result.isPending && currentLastModificationDate.isAfter(cacheEntry.timestamp))
+        ) {
+            logger.info { "InfoResult for $path is either missing, or stale. Checking again" }
+
+            val newResult = InfoResult(asyncPromise, currentLastModificationDate)
+            infoCache[path] = newResult
+            newResult
         } else {
-            cacheResult
+            cacheEntry
         }
-        infoCache[path] = infoResult
-
-        return infoResult.result
     }
 
-    abstract fun getFileInfo(path: String): T
+    @Suppress("UNCHECKED_CAST")
```

Comment:
Is this unchecked cast okay?

Reply:
It could fail if promise was cancelled/interrupted. Added a check to make sure if was successful